device/core/src/client_diagnostic.ts

```diff
@@ -0,0 +1,205 @@
+   *
+   * @returns {string}
+   */
+  getDiagnosticCreationTimeUtc(): string {
+    return this.diagnosticCreationTimeUtc;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticPropertyData.setDiagnosticCreationTimeUtc
+   * @description       set the diagnostic creation timestamp.
+   * @param {string}    diagnosticId   The value of diagnostic creation timestamp.
+   * @throws {ArgumentError}    If value is invalid
+   */
+  setDiagnosticCreationTimeUtc(diagnosticCreationTimeUtc: string): void {
+    /* Codes_SRS_DIAGNOSTICPROPERTYDATA_01_004: [A valid diagnosticCreationTimeUtc shall not be null or empty.] */
+    if (!diagnosticCreationTimeUtc) {
+      throw new errors.ArgumentError('The diagnosticCreationTimeUtc cannot be null or empty.');
+    }
+    this.diagnosticCreationTimeUtc = diagnosticCreationTimeUtc;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticPropertyData.getCorrelationContext
+   * @description       get the diagnostic correlation context..
+   *
+   * @returns {string}
+   */
+  /* Codes_SRS_DIAGNOSTICPROPERTYDATA_01_005: [The function shall return concat string of all correlation contexts.] */
+  public getCorrelationContext(): string {
+    return DiagnosticPropertyData.DIAGNOSTIC_CONTEXT_CREATION_TIME_UTC_PROPERTY + DiagnosticPropertyData.DIAGNOSTIC_CONTEXT_SYMBOL_EQUAL + this.diagnosticCreationTimeUtc;
+  }
+}
+
+/**
+ * Handling device client diagnostic information.
+ */
+export class DiagnosticClient {
+  // The base of diagnostic id character. 0-9a-zA-Z is 62.
+  static DIAGNOSTIC_ID_CHARACTER_BASE: number = 62;
+
+  private diagSamplingPercentage: number;
+  private currentMessageNumber: number;
+
+  /**
+   * @constructor
+   */
+  constructor() {
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_001: [This constructor shall set sampling percentage to 0.]
+    this.diagSamplingPercentage = 0;
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_002: [This constructor shall set message number to 0.]
+    this.currentMessageNumber = 0;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.getDiagSamplingPercentage
+   * @description       get the diagnostic sampling percentage.
+   *
+   * @returns {number}
+   */
+  getDiagSamplingPercentage(): number {
+    return this.diagSamplingPercentage;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.setDiagSamplingPercentage
+   * @description       set the diagnostic sampling percentage.
+   *
+   * @param {number}   diagSamplingPercentage   The value of sampling percentage.
+   * @throws {ArgumentError}     If value < 0 or value > 100.
+   */
+  setDiagSamplingPercentage(diagSamplingPercentage: number): void {
+    if (typeof diagSamplingPercentage !== 'number') {
+      throw new errors.ArgumentError('Sampling percentage only accept number');
+    }
+    if (diagSamplingPercentage % 1 !== 0) {
+      throw new errors.ArgumentError('Sampling percentage should be integer');
+    }
```

Comment:
do we care if it's not? I wouldn't think so.

Reply:
Since our work around in phase 1 is to use device twin properties to deliver the setting from cloud to device, and we only allow integer from 0 - 100(float number is not necessary to support here). We would better check if the value is legal here since the input value is fetched from device twin and user can set any value to it.

Reply:
understood but in case of wrong value, don't throw. that would crash the whole device client. (see my other comment)

---
In reply to: [155440982](https://github.com/Azure/azure-iot-sdk-node/pull/158#discussion_r155440982) [](ancestors = 155440982)